scene/gui/popup.cpp

```diff
@@ -239,44 +256,146 @@ Size2 PopupPanel::_get_contents_minimum_size() const {
+		return;
+	}
+
+	const Ref<StyleBoxFlat> sb = theme_cache.panel_style;
+	if (sb.is_null()) {
+		panel->set_offsets_preset(Control::PRESET_FULL_RECT, Control::PRESET_MODE_MINSIZE, 0);
+		return;
+	}
+
+	const int shadow_size = sb->get_shadow_size();
+	if (shadow_size == 0) {
+		panel->set_offsets_preset(Control::PRESET_FULL_RECT, Control::PRESET_MODE_MINSIZE, 0);
+		return;
+	}
+
+	// Offset the background panel so it leaves space inside the window for the shadows to be drawn.
+	const Point2 shadow_offset = sb->get_shadow_offset();
+	if (is_layout_rtl()) {
+		panel->set_offset(SIDE_LEFT, shadow_size + shadow_offset.x);
+		panel->set_offset(SIDE_RIGHT, -shadow_size + shadow_offset.x);
+	} else {
+		panel->set_offset(SIDE_LEFT, shadow_size - shadow_offset.x);
+		panel->set_offset(SIDE_RIGHT, -shadow_size - shadow_offset.x);
+	}
+	panel->set_offset(SIDE_TOP, shadow_size - shadow_offset.y);
+	panel->set_offset(SIDE_BOTTOM, -shadow_size - shadow_offset.y);
+}
+
+void PopupPanel::_update_child_rects() const {
 	Vector2 cpos(theme_cache.panel_style->get_offset());
-	Vector2 panel_size = Vector2(get_size()) / get_content_scale_factor();
-	Vector2 csize = panel_size - theme_cache.panel_style->get_minimum_size();
+	cpos += Vector2(is_layout_rtl() ? -panel->get_offset(SIDE_RIGHT) : panel->get_offset(SIDE_LEFT), panel->get_offset(SIDE_TOP));
+
+	Vector2 csize = Vector2(get_size()) / get_content_scale_factor() - theme_cache.panel_style->get_minimum_size();
+	// Trim shadows.
+	csize.width -= panel->get_offset(SIDE_LEFT) - panel->get_offset(SIDE_RIGHT);
+	csize.height -= panel->get_offset(SIDE_TOP) - panel->get_offset(SIDE_BOTTOM);
 
 	for (int i = 0; i < get_child_count(); i++) {
 		Control *c = Object::cast_to<Control>(get_child(i));
-		if (!c) {
+		if (!c || c == panel) {
 			continue;
 		}
 
 		if (c->is_set_as_top_level()) {
 			continue;
 		}
 
-		if (c == panel) {
-			c->set_position(Vector2());
-			c->set_size(panel_size);
-		} else {
-			c->set_position(cpos);
-			c->set_size(csize);
-		}
+		c->set_position(cpos);
+		c->set_size(csize);
 	}
 }
 
 void PopupPanel::_notification(int p_what) {
 	switch (p_what) {
-		case NOTIFICATION_READY:
+		case NOTIFICATION_ENTER_TREE: {
+			if (!Engine::get_singleton()->is_editor_hint() && !DisplayServer::get_singleton()->is_window_transparency_available() && !is_embedded()) {
+				Ref<StyleBoxFlat> sb = theme_cache.panel_style;
+				if (sb.is_valid() && (sb->get_shadow_size() > 0 || sb->get_corner_radius(CORNER_TOP_LEFT) > 0 || sb->get_corner_radius(CORNER_TOP_RIGHT) > 0 || sb->get_corner_radius(CORNER_BOTTOM_LEFT) > 0 || sb->get_corner_radius(CORNER_BOTTOM_RIGHT) > 0)) {
+					WARN_PRINT_ONCE("The current theme styles PopupPanel to have shadows and/or rounded corners, but those won't display correctly if 'display/window/per_pixel_transparency/allowed' isn't enabled in the Project Settings, nor if it isn't supported.");
+				}
+			}
```

Comment:
This displays in the Project Manager (which doesn't satisfy `is_editor_hint()`).


```
WARNING: The current theme styles PopupMenu to have shadows and/or rounded corners, but those won't display correctly if 'display/window/per_pixel_transparency/allowed' isn't enabled in the Project Settings, nor if it isn't supported.
     at: _notification (./scene/gui/popup_menu.cpp:1113)
WARNING: The current theme styles PopupPanel to have shadows and/or rounded corners, but those won't display correctly if 'display/window/per_pixel_transparency/allowed' isn't enabled in the Project Settings, nor if it isn't supported.
     at: _notification (./scene/gui/popup.cpp:348)
```